projects/incubator/src/eventsystem/events/event.emitter.ts

```diff
@@ -0,0 +1,97 @@
+import { Injectable, Type } from '@angular/core';
+import { BehaviorSubject, combineLatest, merge, Observable, of } from 'rxjs';
+import { map, switchMap } from 'rxjs/operators';
+import { CxEventSource } from './event.model';
+
+/**
+ * The EventEmitter provides a service to control events in Spartacus.
+ *
+ * Events are driven by an Event Type. Event types are global Classes
+ * which might not be in the final build, this depends on the modules
+ * you use.
+ *
+ * Event sources can be attach and dispatch with the `EventEmiter`.
+ * The various sources are combined into a single emission, so that a subscriber
+ * will get the events in a combined observable.
+ */
+@Injectable({
+  providedIn: 'root',
+})
+export class EventEmitter {
+  /**
+   * The various event sources are collected in a map, stored by
+   * the `EventType` key. The event sources are organized in the
+   * `CxEventSource`, so that additional (runtime) data might be
+   * stored in the future.
+   *
+   * This is a private structure, which might evolve over time.
+   */
+  private events = new Map<Type<any>, CxEventSource<any>>();
+
+  /**
+   * Attach an event source for the given Event Type.
+   *
+   * @param eventType the (global) event type
+   * @param source an observable that represents the source
+   */
+  attach<T>(eventType: Type<T>, source: Observable<T>): void {
+    const event = this.getEvent(eventType);
+    const sources: Observable<T>[] = event.sources.value;
+    sources.push(source);
```

Comment:
@dunqan what's the use case for `distinctUntilChanged`?

I'd like to avoid the re-creation of the array on each `attach()` call for a certain event type due to performance reasons. For _n_ events it's _n(n+1)/2 = O(n^2)_. If the number of events is small, we can accept that, but only when justified.